fe1-web/src/features/lao/reducer/ServerReducer.ts

```diff
@@ -0,0 +1,198 @@
+/**
+ * This error is disabled since reducers use the createSlice function, which requires the user to
+ * param-reassign. Please do not disable other errors.
+ */
+/* eslint-disable no-param-reassign */
+import { createSelector, createSlice, Draft, PayloadAction } from '@reduxjs/toolkit';
+
+import { PublicKey } from 'core/objects';
+
+import { Server, ServerAddress, ServerState } from '../objects/Server';
+import { getLaosState } from './LaoReducer';
+
+/**
+ * Reducer & associated function implementation to store all known servers
+ */
+
+export interface ServerReducerState {
+  byLaoId: {
+    [laoId: string]: {
+      byAddress: {
+        [address: ServerAddress]: ServerState;
+      };
+      allAddresses: ServerAddress[];
+      backendKeyByFrontendKey: {
+        [frontendKey: string]: string;
+      };
+    };
+  };
+}
+
+const initialState: ServerReducerState = {
+  byLaoId: {},
+};
+
+export const SERVER_REDUCER_PATH = 'servers';
+
+const serverSlice = createSlice({
+  name: SERVER_REDUCER_PATH,
+  initialState,
+  reducers: {
+    addServer: (state: Draft<ServerReducerState>, action: PayloadAction<ServerState>) => {
+      const server = action.payload;
+
+      let laoState: ServerReducerState['byLaoId']['string'] = {
+        byAddress: {},
+        allAddresses: [],
+        backendKeyByFrontendKey: {},
+      };
```

Comment:
I'm not sure to understand why this object is created empty like this. Wouldn't it be clearer if it was created at the end with the correct data directly? Plus I admit I don't understand why is there this `['string']`.

Reply:
I did this because the `byLao` can either by of this type or `undefined` and I wanted to avoid writing similar code twice 😅  The next lines check whether it is undefined or not. If it is not undefined, `laoState = state.byLaoId[server.laoId];` is executed and after this line we either have the current laoState (if there was one previously) or we have a new one (what is initialized in the highlighted lines)

This then allows me to simply set `state.byLaoId[server.laoId] = laoState;` after changing the variable `laoState` in this function.

This `['string']` thing is very ugly I agree! This is just for deriving the type for the values in the `byLaoId` object/map. I changed it to be more verbose but hopefully easier to understand :)

Reply:
Okay, thanks for the explanation! It's indeed easier to understand with your changes imo.